src/data/ellpack_page_source.cu

```diff
@@ -1,29 +1,161 @@
 /**
  * Copyright 2019-2024, XGBoost contributors
  */
-#include <memory>
+#include <thrust/host_vector.h>  // for host_vector
 
-#include "ellpack_page.cuh"
-#include "ellpack_page.h"  // for EllpackPage
+#include <cstddef>  // for size_t
+#include <cstdint>  // for int8_t, uint64_t, uint32_t
+#include <memory>   // for shared_ptr, make_unique, make_shared
+#include <utility>  // for move
+
+#include "../common/common.h"                 // for safe_cuda
+#include "../common/cuda_pinned_allocator.h"  // for pinned_allocator
+#include "../common/device_helpers.cuh"       // for CUDAStreamView, DefaultStream
+#include "ellpack_page.cuh"                   // for EllpackPageImpl
+#include "ellpack_page.h"                     // for EllpackPage
 #include "ellpack_page_source.h"
+#include "xgboost/base.h"  // for bst_idx_t
 
 namespace xgboost::data {
-void EllpackPageSource::Fetch() {
-  dh::safe_cuda(cudaSetDevice(device_.ordinal));
+struct EllpackHostCache {
+  thrust::host_vector<std::int8_t, common::cuda::pinned_allocator<std::int8_t>> cache;
+
+  void Resize(std::size_t n, dh::CUDAStreamView stream) {
+    stream.Sync();  // Prevent partial copy inside resize.
+    cache.resize(n);
+  }
+};
+
+class EllpackHostCacheStreamImpl {
+  std::shared_ptr<EllpackHostCache> cache_;
+  bst_idx_t cur_ptr_{0};
+  bst_idx_t bound_{0};
+
+ public:
+  explicit EllpackHostCacheStreamImpl(std::shared_ptr<EllpackHostCache> cache)
+      : cache_{std::move(cache)} {}
+
+  [[nodiscard]] bst_idx_t Write(void const* ptr, bst_idx_t n_bytes) {
+    auto n = cur_ptr_ + n_bytes;
+    if (n > cache_->cache.size()) {
+      cache_->Resize(n, dh::DefaultStream());
```

Comment:
Allocating pinned memory seems pretty expensive. I wonder if it's worth calculating the total size and allocate the cache in one shot.

Reply:
That requires an additional iteration over the disk cache, which is not cheap either.

Reply:
The vector class uses dynamic table internally, the amortized cost should be reasonable, we can look into the initialization cost after removing the disk cache entirely.

Reply:
Alternatively, as you suggested, we can let system pagable memory to back the main cache.

Reply:
In my test, allocating a 120GB pinned memory buffer takes 5-6 seconds. But as a first pass, this is probably fine.

Reply:
It's expensive indeed. If we extrapolate on that speed assuming it's linear, then 500 GB would be 21 seconds, probably fine for the initial implementation. We will look into malloc as the main cache.